modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -127,15 +345,36 @@ void HashTSDFVolumeCPU::integrate(InputArray _depth, float depthFactor, const Ma
     parallel_for_(allocateRange, AllocateVolumeUnitsInvoker);
 
     //! Perform the allocation
-    int res = volumeUnitResolution;
-    Point3i volumeDims(res, res, res);
     for (auto idx : newIndices)
     {
         VolumeUnit& vu = volumeUnits[idx];
         Matx44f subvolumePose = pose.translate(volumeUnitIdxToVolume(idx)).matrix;
-        vu.pVolume = makePtr<TSDFVolumeCPU>(voxelSize, subvolumePose, raycastStepFactor, truncDist, maxWeight, volumeDims);
+        
+        int xdim, ydim, zdim;
+        if (zFirstMemOrder)
+        {
+            xdim = volumeDims.z * volumeDims.y;
+            ydim = volumeDims.z;
+            zdim = 1;
+        }
+        else
+        {
+            xdim = 1;
+            ydim = volumeDims.x;
+            zdim = volumeDims.x * volumeDims.y;
+        }
+        vu.volDims = Vec4i(xdim, ydim, zdim);
```

Comment:
`volDims` can be precalculated in the class constructor

Reply:
done